consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java

```diff
@@ -0,0 +1,137 @@
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Responsible for handling all RoundChange messages received for a given block height
+ * (theoretically, RoundChange messages for a older height should have been previously discarded,
+ * and messages for a future round should have been buffered).
+ *
+ * <p>If enough RoundChange messages all targeting a given round are received (and this node is the
+ * proposer for said round) - a newRound message is sent, and a new round should be started by the
+ * controlling class.
+ */
+public class RoundChangeManager {
+
+  public static class RoundChangeStatus {
+    private final int quorumSize;
+
+    @VisibleForTesting
+    final Map<Address, SignedData<RoundChangePayload>> receivedMessages = Maps.newHashMap();
+
+    private boolean actioned = false;
+
+    public RoundChangeStatus(final int quorumSize) {
+      this.quorumSize = quorumSize;
+    }
+
+    public void addMessage(final SignedData<RoundChangePayload> msg) {
+      if (!actioned) {
+        receivedMessages.put(msg.getSender(), msg);
+      }
+    }
+
+    public boolean roundChangeReady() {
+      return receivedMessages.size() >= quorumSize && !actioned;
+    }
+
+    public RoundChangeCertificate createRoundChangeCertificate() {
+      if (roundChangeReady()) {
+        actioned = true;
+        return new RoundChangeCertificate(receivedMessages.values());
+      } else {
+        throw new IllegalStateException("Unable to create RoundChangeCertificate at this time.");
+      }
+    }
+  }
+
+  private static final Logger LOG = LogManager.getLogger();
+
+  @VisibleForTesting
+  final Map<ConsensusRoundIdentifier, RoundChangeStatus> roundChangeCache = Maps.newHashMap();
+
+  private final Collection<Address> validators;
+  private final int quorumSize;
+  private final long sequenceNumber;
+  private final MessageValidatorFactory messageValidityFactory;
+
+  public RoundChangeManager(
+      final long sequenceNumber,
+      final Collection<Address> validators,
+      final MessageValidatorFactory messageValidityFactory) {
+    this.validators = validators;
+    this.messageValidityFactory = messageValidityFactory;
+    this.sequenceNumber = sequenceNumber;
+    this.quorumSize = IbftHelpers.calculateRequiredValidatorQuorum(validators.size());
+  }
+
+  public Optional<RoundChangeCertificate> appendRoundChangeMessage(
+      final SignedData<RoundChangePayload> msg) {
+    // validate the received msg
+    final ConsensusRoundIdentifier msgTargetRound = msg.getPayload().getRoundChangeIdentifier();
+    final RoundChangeMessageValidator messageValidator =
+        new RoundChangeMessageValidator(
+            messageValidityFactory, validators, quorumSize, sequenceNumber);
+    if (!messageValidator.validateMessage(msg)) {
+      LOG.info("RoundChange message was invalid.");
```

Comment:
I think the WARN level makes more sense than info as this is potentially an problem something should look at